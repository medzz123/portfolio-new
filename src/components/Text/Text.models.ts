import { Fonts, FontSizes, FontWeights } from '@theme/font';
import { Palette } from '@theme/palette';
import { ReactNode } from 'react';

export interface TextProps {
  /**
   * Type of element
   */
  variant?: TextVariant;
  /**
   * Replaces default variant
   */
  size?: FontSizes;
  /**
   * Font weight
   */
  weight?: FontWeights;
  /**
   * Color
   */
  color?: Palette;
  /**
   * Text Align
   */
  align?: 'left' | 'right' | 'center';
  /**
   * Margin Bottom
   */
  mb?: SpacingVariants;
  /**
   * Font family
   */
  font?: Fonts;
  children?: ReactNode;
  className?: string;
}

export type SpacingVariants = 'none' | 'small' | 'large';

export type TextVariant = 'p' | 'span' | 'h1' | 'h2' | 'h3';
