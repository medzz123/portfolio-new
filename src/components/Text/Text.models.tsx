import { FontSizes, FontWeights } from '@theme/font';

export interface TextProps {
  /**
   * Type of element
   */
  variant?: TextVariant;
  /**
   * Replaces default variant
   */
  level?: FontSizes;
  /**
   * Font weight
   */
  weight?: FontWeights;
  /**
   * Color
   */
  color?: 'error' | 'success' | 'normal' | 'info';
  /**
   * Text Align
   */
  align?: 'left' | 'right' | 'center';
  /**
   * Margin Bottom
   */
  mb?: 'none' | 'small' | 'large';
}

export type TextVariant = 'p' | 'span' | 'h1' | 'h2' | 'h3';
