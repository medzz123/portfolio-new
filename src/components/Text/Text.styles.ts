import { styled } from '@theme/config';

export const TextContainer = styled('p', {
  margin: '$0',
  lineHeight: '1.5',
  variants: {
    level: {
      normal: {
        fontSize: '$normal',
      },
      small: {
        fontSize: '$small',
      },
      info: {
        fontSize: '$info',
      },
      h1: {
        fontSize: '$h1',
        lineHeight: '1',
      },
      h2: {
        fontSize: '$h2',
        lineHeight: '1.3',
      },
      h3: {
        fontSize: '$h3',
        lineHeight: '1.3',
      },
    },
    weight: {
      thin: {
        fontWeight: '$thin',
      },
      normal: {
        fontWeight: '$normal',
      },
      semiBold: {
        fontWeight: '$semiBold',
      },
      bold: {
        fontWeight: '$bold',
      },
      black: { fontWeight: '$black' },
    },
    color: {
      success: {
        color: '$hiContrast',
      },
      error: {
        color: '$error',
      },
      info: {
        color: '$hiContrast',
      },
      normal: {
        color: '$hiContrast',
      },
    },
    align: {
      left: {
        textAlign: 'left',
      },
      right: {
        textAlign: 'right',
      },
      center: {
        textAlign: 'center',
      },
    },
    spacing: {
      none: {
        marginBottom: '$0',
      },
      small: {
        marginBottom: '$20',
      },
      large: {
        marginBottom: '$30',
      },
    },
    font: {
      abril: {
        fontFamily: '$abril',
      },
      mulish: {
        fontFamily: '$mulish',
      },
    },
  },
});
