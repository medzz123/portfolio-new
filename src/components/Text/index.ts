export { default } from './Text';
export { H1, H2, H3, P, S } from './Text.variants';
