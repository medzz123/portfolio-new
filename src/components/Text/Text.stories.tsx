import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';

import Text from './Text';

export default {
  title: 'Components/Text',
  component: Text,
} as Meta;

const Template: Story = (storyArguments) => <Text {...storyArguments}>Hello</Text>;

export const Default = Template.bind({});
