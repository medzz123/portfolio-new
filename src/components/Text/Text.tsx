import React, { FunctionComponent } from 'react';

import { TextProps } from './Text.models';
import { TextContainer } from './Text.styles';
import { mapSize } from './Text.utils';

const Text: FunctionComponent<TextProps> = (props) => {
  const {
    children,
    level,
    weight = 'normal',
    color = 'normal',
    mb = 'small',
    align = 'left',
    variant = 'p',
  } = props;

  const mappedLevel = level || mapSize[variant];

  return (
    <TextContainer
      as={variant}
      spacing={mb}
      align={align}
      weight={weight}
      color={color}
      level={mappedLevel}
      data-testid="text"
    >
      {children}
    </TextContainer>
  );
};

export default Text;
