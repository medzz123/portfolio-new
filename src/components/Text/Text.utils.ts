import { FontSizes } from '@theme/font';

import { TextVariant } from './Text.models';

export const mapSize: {
  [key in TextVariant]: FontSizes;
} = {
  p: 'normal',
  span: 'normal',
  h1: 'h1',
  h2: 'h2',
  h3: 'h3',
};
