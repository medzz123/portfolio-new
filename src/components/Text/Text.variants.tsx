import { forwardRef } from 'react';

import Text from './Text';
import { TextProps } from './Text.models';

const H1 = forwardRef<HTMLHeadingElement, Omit<TextProps, 'variant'>>(
  (props, ref) => (
    <Text ref={ref} variant="h1" {...props}>
      {props?.children}
    </Text>
  )
);

const H2 = forwardRef<HTMLHeadingElement, Omit<TextProps, 'variant'>>(
  (props, ref) => (
    <Text ref={ref} variant="h2" {...props}>
      {props?.children}
    </Text>
  )
);

const H3 = forwardRef<HTMLHeadingElement, Omit<TextProps, 'variant'>>(
  (props, ref) => (
    <Text ref={ref} variant="h3" {...props}>
      {props?.children}
    </Text>
  )
);

const P = forwardRef<HTMLParagraphElement, Omit<TextProps, 'variant'>>(
  (props, ref) => (
    <Text ref={ref} variant="p" {...props}>
      {props?.children}
    </Text>
  )
);

const S = forwardRef<HTMLParagraphElement, Omit<TextProps, 'variant'>>(
  (props, ref) => (
    <Text ref={ref} variant="span" {...props}>
      {props?.children}
    </Text>
  )
);

export { H1, H2, H3, P, S };
