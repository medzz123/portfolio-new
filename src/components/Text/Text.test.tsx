import { render } from '@test/library';
import React from 'react';

import Text from './Text';

test('Text Test', () => {
  const { getByTestId, container } = render(<Text>Hello</Text>);

  const textNode = getByTestId('text');

  expect(textNode.textContent).toBe('Hello');

  expect(container).toMatchSnapshot();
});
