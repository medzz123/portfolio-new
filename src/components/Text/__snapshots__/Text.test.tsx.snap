// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Test 1`] = `
<div>
  <p
    class="sx3wn9e sx3wn9e9n3hc--level-normal sx3wn9eh4u4f--weight-normal sx3wn9efxkwo--color-normal sx3wn9e0o7ux--align-left sx3wn9ep1cgo--spacing-small"
    data-testid="text"
  >
    Hello
  </p>
</div>
`;
